File: src/test/java/agh/ics/oop/model/Vector2dTest.java
package agh.ics.oop.model;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

public class Vector2dTest {
    @Test
    public void testEquals() {
        assert(new Vector2d(1, 2).equals(new Vector2d(1, 2)));
        assert(!new Vector2d(1, 2).equals(new Vector2d(1, 3)));
        assert(!new Vector2d(1, 2).equals(new Vector2d(2, 2)));
        assert(!new Vector2d(1, 2).equals(new Vector2d(2, 3)));
    }

    @Test
    public void testToString() {
        assertEquals(new Vector2d(1, 2).toString(), "(1,2)");
        assertEquals(new Vector2d(1, 3).toString(), "(1,3)");
        assertEquals(new Vector2d(2, 2).toString(), "(2,2)");
        assertNotEquals(new Vector2d(2, 3).toString(), "(2,2)");
        assertNotEquals(new Vector2d(2, 3).toString(), "(1,3)");
    }

    @Test
    public void testPrecedes() {
        assert(new Vector2d(1, 2).precedes(new Vector2d(1, 2)));
        assert(new Vector2d(1, 2).precedes(new Vector2d(1, 3)));
        assert(new Vector2d(1, 4).precedes(new Vector2d(2, 8)));
        assert(new Vector2d(1, 2).precedes(new Vector2d(2, 3)));
        assert(!new Vector2d(1, 2).precedes(new Vector2d(1, 1)));
        assert(new Vector2d(3, 2).precedes(new Vector2d(10, 2)));
        assert(!new Vector2d(1, 2).precedes(new Vector2d(0, 1)));
    }

    @Test
    public void testFollows() {
        assert(new Vector2d(1, 2).follows(new Vector2d(1, 2)));
        assert(new Vector2d(1, 3).follows(new Vector2d(1, 2)));
        assert(new Vector2d(2, 8).follows(new Vector2d(1, 4)));
        assert(new Vector2d(2, 3).follows(new Vector2d(1, 2)));
        assert(!new Vector2d(1, 1).follows(new Vector2d(1, 2)));
        assert(new Vector2d(10, 2).follows(new Vector2d(3, 2)));
        assert(!new Vector2d(0, 1).follows(new Vector2d(1, 2)));
    }

    @Test
    public void testUpperRight() {
        assertEquals(new Vector2d(1, 2).upperRight(new Vector2d(1, 2)), new Vector2d(1, 2));
        assertEquals(new Vector2d(1, 2).upperRight(new Vector2d(1, 3)), new Vector2d(1, 3));
        assertEquals(new Vector2d(1, 4).upperRight(new Vector2d(2, 8)), new Vector2d(2, 8));
        assertEquals(new Vector2d(1, 2).upperRight(new Vector2d(2, 3)), new Vector2d(2, 3));
        assertEquals(new Vector2d(1, 2).upperRight(new Vector2d(1, 1)), new Vector2d(1, 2));
        assertEquals(new Vector2d(3, 2).upperRight(new Vector2d(10, 2)), new Vector2d(10, 2));
        assertEquals(new Vector2d(1, 2).upperRight(new Vector2d(0, 1)), new Vector2d(1, 2));
    }

    @Test
    public void testLowerLeft() {
        assertEquals(new Vector2d(1, 2).lowerLeft(new Vector2d(1, 2)), new Vector2d(1, 2));
        assertEquals(new Vector2d(1, 2).lowerLeft(new Vector2d(1, 3)), new Vector2d(1, 2));
        assertEquals(new Vector2d(1, 4).lowerLeft(new Vector2d(2, 8)), new Vector2d(1, 4));
        assertEquals(new Vector2d(1, 2).lowerLeft(new Vector2d(2, 3)), new Vector2d(1, 2));
        assertEquals(new Vector2d(1, 2).lowerLeft(new Vector2d(1, 1)), new Vector2d(1, 1));
        assertEquals(new Vector2d(3, 2).lowerLeft(new Vector2d(10, 2)), new Vector2d(3, 2));
        assertEquals(new Vector2d(1, 2).lowerLeft(new Vector2d(0, 1)), new Vector2d(0, 1));
    }

    @Test
    public void testAdd() {
        assertEquals(new Vector2d(1, 2).add(new Vector2d(1, 2)), new Vector2d(2, 4));
        assertEquals(new Vector2d(1, 2).add(new Vector2d(1, 3)), new Vector2d(2, 5));
        assertEquals(new Vector2d(1, 4).add(new Vector2d(2, 8)), new Vector2d(3, 12));
        assertEquals(new Vector2d(1, 2).add(new Vector2d(2, 3)), new Vector2d(3, 5));
        assertEquals(new Vector2d(1, 2).add(new Vector2d(1, 1)), new Vector2d(2, 3));
        assertEquals(new Vector2d(3, 2).add(new Vector2d(10, 2)), new Vector2d(13, 4));
        assertEquals(new Vector2d(1, 2).add(new Vector2d(0, 1)), new Vector2d(1, 3));
    }

    @Test
    public void testSubtract() {
        assertEquals(new Vector2d(1, 2).subtract(new Vector2d(1, 2)), new Vector2d(0, 0));
        assertEquals(new Vector2d(1, 2).subtract(new Vector2d(1, 3)), new Vector2d(0, -1));
        assertEquals(new Vector2d(1, 4).subtract(new Vector2d(2, 8)), new Vector2d(-1, -4));
        assertEquals(new Vector2d(1, 2).subtract(new Vector2d(2, 3)), new Vector2d(-1, -1));
        assertEquals(new Vector2d(1, 2).subtract(new Vector2d(1, 1)), new Vector2d(0, 1));
        assertEquals(new Vector2d(3, 2).subtract(new Vector2d(10, 2)), new Vector2d(-7, 0));
        assertEquals(new Vector2d(1, 2).subtract(new Vector2d(0, 1)), new Vector2d(1, 1));
    }

    @Test
    public void testOpposite() {
        assertEquals(new Vector2d(1, 2).opposite(), new Vector2d(-1, -2));
        assertEquals(new Vector2d(1, 3).opposite(), new Vector2d(-1, -3));
        assertEquals(new Vector2d(2, 2).opposite(), new Vector2d(-2, -2));
        assertEquals(new Vector2d(2, 3).opposite(), new Vector2d(-2, -3));
        assertEquals(new Vector2d(1, 2).opposite(), new Vector2d(-1, -2));
        assertEquals(new Vector2d(3, 2).opposite(), new Vector2d(-3, -2));
        assertEquals(new Vector2d(1, 2).opposite(), new Vector2d(-1, -2));
    }
}

File: kotlin/src/test/kotlin/model/Vector2dTest.kt
package model

import org.junit.jupiter.api.Assertions.assertEquals
import kotlin.test.Test

class Vector2dTest {
    @Test
    fun `follows should return true when vector follows the other vector and false when not`() {
        val vector1 = Vector2d(3, 5)
        val vector2 = Vector2d(2, 4)
        val vector3 = Vector2d(3, 5)
        val vector4 = Vector2d(2, 6)

        assertEquals(true, vector1 follows vector2)
        assertEquals(true, vector1 follows vector3)
        assertEquals(false, vector2 follows vector1)
        assertEquals(false, vector3 follows vector4)
    }

    @Test
    fun `precedes should return true when vector precedes the other vector and false when not`() {
        val vector1 = Vector2d(3, 5)
        val vector2 = Vector2d(4, 6)
        val vector3 = Vector2d(3, 5)
        val vector4 = Vector2d(2, 6)

        assertEquals(true, vector1 precedes vector2)
        assertEquals(true, vector1 precedes vector3)
        assertEquals(false, vector2 precedes vector1)
        assertEquals(false, vector2 precedes vector4)
    }

    @Test
    fun `upperRight should return the upper right vector`() {
        val vector1 = Vector2d(3, 5)
        val vector2 = Vector2d(2, 8)
        val vector3 = Vector2d(4, 6)

        assertEquals(Vector2d(3, 8), vector1.upperRight(vector2))
        assertEquals(Vector2d(4, 6), vector1.upperRight(vector3))
    }

    @Test
    fun `lowerLeft should return the lower left vector`() {
        val vector1 = Vector2d(3, 5)
        val vector2 = Vector2d(2, 4)
        val vector3 = Vector2d(4, 6)

        assertEquals(Vector2d(2, 4), vector1.lowerLeft(vector2))
        assertEquals(Vector2d(3, 5), vector1.lowerLeft(vector3))
    }

    @Test
    fun `opposite should return the vector with swapped coordinates`() {
        val vector = Vector2d(3, 5)

        assertEquals(Vector2d(5, 3), vector.opposite())
    }

    @Test
    fun `toString should return the string representation of the vector`() {
        val vector = Vector2d(3, 5)

        assertEquals("(3,5)", vector.toString())
    }
    @Test
    fun `add should return the sum of two vectors`() {
        val vector1 = Vector2d(3, 5)
        val vector2 = Vector2d(2, 4)

        assertEquals(Vector2d(5, 9), vector1 + vector2)
    }
    @Test
    fun `subtract should return the difference of two vectors`() {
        val vector1 = Vector2d(3, 5)
        val vector2 = Vector2d(2, 4)

        assertEquals(Vector2d(1, 1), vector1 - vector2)
    }
    @Test
    fun `equaling should return if two vectors are equal`() {
        val vector1 = Vector2d(3, 5)
        val vector2 = Vector2d(2, 4)
        val vector3 = Vector2d(3, 5)

        assertEquals(false, vector1 == vector2)
        assertEquals(true, vector1 == vector3)
    }
    @Test
    fun `toUnitVector should return the correct unit vector`() {
        assertEquals(Vector2d(0, 1), MapDirection.NORTH.toUnitVector())
        assertEquals(Vector2d(-1, 0), MapDirection.WEST.toUnitVector())
        assertEquals(Vector2d(0, -1), MapDirection.SOUTH.toUnitVector())
        assertEquals(Vector2d(1, 0), MapDirection.EAST.toUnitVector())
    }
}